// @usedapp/core/dist/esm/src/hooks/useToken.js
import { ERC20Interface } from '../constants';
import { useCalls } from './useCall';
import { Contract } from 'ethers';
/**
 * @public
 */
export function useToken(tokenAddress) {
    var _a, _b;
    const partialCall = {
        contract: new Contract(tokenAddress || '', ERC20Interface),
        address: tokenAddress || '',
        args: [],
    };
    const args = ['name', 'symbol', 'decimals', 'totalSupply'].map((method) => (Object.assign(Object.assign({}, partialCall), { method })));
    const [name, symbol, decimals, totalSupply] = useCalls(args);
    if (!name && !symbol && !decimals && !totalSupply) {
        return undefined;
    }
    return {
        name: (_a = name === null || name === void 0 ? void 0 : name.value[0]) !== null && _a !== void 0 ? _a : '',
        symbol: (_b = symbol === null || symbol === void 0 ? void 0 : symbol.value[0]) !== null && _b !== void 0 ? _b : '',
        decimals: decimals === null || decimals === void 0 ? void 0 : decimals.value[0],
        totalSupply: totalSupply === null || totalSupply === void 0 ? void 0 : totalSupply.value[0],
    };
}
